src/main/java/de/metas/ui/web/handlingunits/process/WEBUIHUCreationWithSerialNumberService.java

```diff
@@ -0,0 +1,262 @@
+		}
+
+		return WebuiHUTransformCommandResult.builder()
+				.huIdsChanged(huIDsChanged)
+				.huIdsToAddToView(huIDsAdded)
+				.build();
+	}
+
+	private Set<Integer> splitIntoCUs(final HUEditorRow.HUEditorRowHierarchy huEditorRowHierarchy, final int cuNumber)
+	{
+		if (cuNumber == 0)
+		{
+			return ImmutableSet.of();
+		}
+
+		final Set<Integer> splitCUs = new HashSet<>();
+
+		while (cuNumber > splitCUs.size())
+		{
+			final int maxCUsAllowedPerBatch = cuNumber - splitCUs.size();
+			splitCUs.addAll(createCUsBatch(huEditorRowHierarchy, maxCUsAllowedPerBatch));
+		}
+
+		return splitCUs;
+	}
+
+	private Set<Integer> createCUsBatch(final HUEditorRow.HUEditorRowHierarchy huEditorRowHierarchy, final int maxCUsAllowedPerBatch)
+	{
+		final Set<Integer> splitCUIDs = new HashSet<>();
+
+		final IHandlingUnitsDAO handlingUnitsDAO = Services.get(IHandlingUnitsDAO.class);
+
+		final HUEditorRow cuRow = huEditorRowHierarchy.getCuRow();
+
+		final HUEditorRow parentRow = huEditorRowHierarchy.getParentRow();
+
+		final int initialQtyCU = cuRow.getQtyCU().intValueExact();
+
+		I_M_HU huToSplit = cuRow.getM_HU();
+		int numberOfCUsToCreate;
+
+		if (parentRow == null)
+		{
+			// The CU will not be split when its qty gets to 1. So make sure the selected CU gets in the list of split CUs
+			final int selectedCUID = huToSplit.getM_HU_ID();
+			splitCUIDs.add(selectedCUID);
+
+			numberOfCUsToCreate = maxCUsAllowedPerBatch < initialQtyCU ? maxCUsAllowedPerBatch : initialQtyCU;
+
+			for (int i = 0; i < numberOfCUsToCreate; i++)
+			{
+				final List<I_M_HU> createdCUs = newHUTransformation().cuToNewCU(huToSplit, BigDecimal.ONE);
+				splitCUIDs.addAll(createdCUs.stream().map(I_M_HU::getM_HU_ID).collect(ImmutableSet.toImmutableSet()));
+			}
+		}
+		else
+		{
+			Check.assume(parentRow.isTU(), " Parent row must be TU: " + parentRow);
+			I_M_HU parentHU = parentRow.getM_HU();
+
+			if (isAggregateHU(parentRow))
+			{
+				final HUEditorRow luRow = huEditorRowHierarchy.getTopLevelRow();
+
+				parentHU = createNonAggregatedTU(parentRow, luRow);
+
+				huToSplit = handlingUnitsDAO.retrieveIncludedHUs(parentHU).get(0);
+			}
+
+			final int tuCapacity = calculateTUCapacity(parentHU);
+			numberOfCUsToCreate = getMinimumOfThree(tuCapacity, maxCUsAllowedPerBatch, initialQtyCU);
+			for (int i = 0; i < numberOfCUsToCreate; i++)
+			{
+				final List<I_M_HU> createdCUs = newHUTransformation().cuToExistingTU(huToSplit, BigDecimal.ONE, parentHU);
+
+				splitCUIDs.addAll(createdCUs.stream().map(I_M_HU::getM_HU_ID).collect(ImmutableSet.toImmutableSet()));
+			}
+		}
+
+		return splitCUIDs;
+	}
+
+	private int calculateTUCapacity(final I_M_HU parentHU)
+	{
+		final I_M_HU_PI_Item_Product huPIP = parentHU.getM_HU_PI_Item_Product();
+
+		if (huPIP == null)
+		{
+			// TODO
+			throw new UnsupportedOperationException();
+		}
+
+		return huPIP.getQty().intValueExact();
+	}
+
+	private int getMinimumOfThree(final int no1, final int no2, final int no3)
```

Comment:
such expressions make me nervous, they are over my head..unless you are really confident it , please unit-test

Reply:
Utit test added